rfcs/20200618-headers/README.md

```diff
@@ -0,0 +1,123 @@
+# Changes in the oneDNN header install locations and namespace
+
+## Introduction
+
+New headers structure and namespaces are needed for cross-component and oneAPI
+branding alignment. The request from the oneAPI program is to put oneDNN
+headers under the `oneapi` and also to provide a corresponding namespace. The
+changes will have to be reflected in the
+[spec](https://github.com/oneapi-src/oneAPI-spec/tree/master/source/elements/oneDNN).
+
+The main concern is how to keep backward compatibility with the existing code.
+The secondary concern is maintenance burden.
+
+## Proposal
+
+The following changes are to be implemented in the
+[`dev-v2`](https://github.com/oneapi-src/onednn/tree/dev-v2) branch but will
+eventually be part of master.
+
+### Namespace changes
+
+oneDNN should start providing two namespaces (C++ only): `dnnl::` and
+`oneapi::dnnl::`. Either can be an alias. To minimize the changes, the
+proposal is to alias the long one:
+
+```c++
+namespace dnnl { /* the usual stuff */ }
+
+namespace oneapi { namespace dnnl = ::dnnl; }
+```
+
+The new namespace must be reflected in the
+[spec](https://github.com/oneapi-src/oneAPI-spec/tree/master/source/elements/oneDNN).
+There is no recommendation on whether longer or shorted namespace should be
+used in the spec yet; this is outside of the scope of this RFC.
+
+### Installation paths changes
+
+oneDNN is free to decide on the header structure under `include/oneapi`.
+
+#### Option 1 -- recommended
+
+```
+include
+├── oneapi[/dnnl]
+│   ├── dnnl_types.h
+│   ├── dnnl_config.h
+│   ├── dnnl.hpp
+│   └── dnnl.h
```

Comment:
probably obvious, but could you please also state, how `dnnl_types.hpp` will be included in `dnnl.hpp`:
- `#include "oneapi/dnnl/dnnl_types.hpp"` -- preferable, but will require some play with Cmake & generated files
- `#include "dnnl_types.hpp"` -- also works, but could be "in conflict" with `include/dnnl_types.hpp`, not recommended

Reply:
`#include "dnnl_types.h"` [searches the file's current directory first](https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html), so I don't see how there could be a conflict. As far as my reading of the search order goes, no changes to the header files would be required. If you are OK with this statement, I'll update the RFC.

Reply:
> searches the file's current directory first

yeah, that's why put "in conflict" in quotes.

In my opinion, we should use the full path though. The rationale is:
1. It is clearer what file is meant
2. The possible "real" conflict is auto-generated files. When we build `benchdnn` auto-generated headers are not in the same directories as `dnnl.hpp`, so whether a proxy `include/dnnl_version.h` or real auto-generated `$CMAKE_BUILD_DIR/include[/oneapi/dnnl]/dnnl_version.hpp` is included will depend on `-I` order.

The potential downside is, however, that people will be forced to preserve the structure `oneapi/dnnl/...` as otherwise their copy of `oneapi/dnnl/dnnl.hpp` won't be able to include `dnnl_types.hpp`. Maybe this is actually a good thing...

Reply:
1. Okay. I can certainly change that.

2. I'm not sure I understand this scenario. Can you please elaborate?

Reply:
2. I am wrong :) It doesn't actually matter much. If `dnnl.hpp` will include `dnnl_version.h` w/o path, that would simply mean that during building we will always go through the proxy:
```
/include/oneapi/dnnl/dnnl.hpp # include dnnl_version.h
/include/dnnl_version.h # include oneapi/dnnl/dnnl_version.h
/build/include/oneapi/dnnl/dnnl_version.h
```

That's fine, but probably it would be more straightforward to have:
```
/include/oneapi/dnnl/dnnl.hpp # include oneapi/dnnl/dnnl_version.h
/build/include/oneapi/dnnl/dnnl_version.h
```

But yeah, the behavior should be fine in both cases. Resolving this discussion. Sorry for the confusion,

Reply:
I've changed the text to go with your suggestion. I think its cleaner.

Reply:
Thank you, the document LGTM.